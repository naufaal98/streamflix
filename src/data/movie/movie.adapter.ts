import { Movie, MovieList, Cast, MovieDetail } from 'data/movie/movie.type';
import calculatePriceByRating from 'utils/calculatePriceByRating';

export function movieAdapter(movie: any): Movie {
  return {
    id: movie.id,
    title: movie.title,
    rating: movie.vote_average,
    overview: movie.overview,
    release_date: movie.release_date,
    price: calculatePriceByRating(movie.vote_average),
  };
}

export function movieListAdapter(data: any): MovieList {
  return {
    page: data.page,
    results: data.results.map((movie: any) => movieAdapter(movie)),
  };
}

export function castAdapter(cast: any): Cast {
  return {
    id: cast.id,
    name: cast.name,
    profile_path: cast.profile_path,
  };
}

export function movieDetailAdapter(movie: any, casts: any): MovieDetail {
  return {
    ...movieAdapter(movie),
    duration: movie.runtime,
    casts: casts.map((cast: any) => castAdapter(cast)),
  };
}
