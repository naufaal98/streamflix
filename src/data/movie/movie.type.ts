export interface Movie {
  id: number;
  title: string;
  rating: number;
  overview: string;
  release_date: string;
  poster_path: string;
  backdrop_path: string;
  price: number;
}

export interface MovieList {
  page: number;
  results: Movie[];
}

export interface Cast {
  id: number;
  name: string;
  profile_path: string;
}

export interface MovieDetail extends Movie {
  duration: number;
  casts: Cast[];
}
