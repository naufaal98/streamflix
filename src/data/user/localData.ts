import { User } from 'types/User';

const FREE_BALANCE = 100000;

export const setInitialUserData = () => {
  const user: User = {
    balance: FREE_BALANCE,
    purchased_movies: [],
  };

  localStorage.setItem('user', JSON.stringify(user));
  return user;
};

export const setLocalUserData = (user: User) => {
  try {
    localStorage.setItem('user', JSON.stringify(user));
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
  }
};

export const getLocalUserData = () => {
  try {
    if (!localStorage.getItem('user')) {
      return setInitialUserData();
    }

    const user = JSON.parse(localStorage.getItem('user')!) || [];
    return {
      balance: user.balance,
      purchased_movies: user.purchased_movies || [],
    };
  } catch (e) {
    return {
      balance: 0,
      purchased_movies: [],
    };
  }
};
