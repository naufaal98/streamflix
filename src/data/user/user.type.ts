export interface User {
  balance: number;
  purchased_movies: number[];
}
