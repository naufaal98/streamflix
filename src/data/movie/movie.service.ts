import axios from 'axios';
import { BASE_URL } from 'constant';
import { MovieDetail, MovieList } from 'types/Movie';
import { movieListAdapter, movieDetailAdapter } from './movie.adapter';

const API_KEY = process.env.REACT_APP_API_KEY;

interface MovieListParams {
  page: number;
  language?: string;
  region?: string;
}

export async function getNowPlayingMovies({
  page = 1,
  language,
  region,
}: MovieListParams): Promise<MovieList> {
  const params: MovieListParams = { page };
  if (language) params.language = language;
  if (region) params.region = region;

  const URL = `${BASE_URL}/now_playing`;

  const { data } = await axios.get<any>(URL, {
    params: {
      ...params,
      api_key: API_KEY,
    },
  });

  return movieListAdapter(data);
}

export async function getMovieDetail(id: number): Promise<MovieDetail> {
  const URL_DETAIL = `${BASE_URL}/${id}`;
  const URL_CREDIT = `${BASE_URL}/${id}/credits`;

  const { data: movie } = await axios.get<any>(URL_DETAIL, { params: { api_key: API_KEY } });
  const { data: credit } = await axios.get<any>(URL_CREDIT, { params: { api_key: API_KEY } });

  return movieDetailAdapter(movie, credit.cast);
}

export async function getRecommendedMovies(
  id: number,
  { page = 1, language }: MovieListParams,
): Promise<MovieList> {
  const params: MovieListParams = { page };
  if (language) params.language = language;

  const URL = `${BASE_URL}/${id}/recommendations`;

  const { data } = await axios.get<any>(URL, {
    params: {
      ...params,
      api_key: API_KEY,
    },
  });

  return movieListAdapter(data);
}

export async function getSimilarMovies(
  id: number,
  { page = 1, language }: MovieListParams,
): Promise<MovieList> {
  const params: MovieListParams = { page };
  if (language) params.language = language;

  const URL = `${BASE_URL}/${id}/similar`;

  const { data } = await axios.get<any>(URL, {
    params: {
      ...params,
      api_key: API_KEY,
    },
  });

  return movieListAdapter(data);
}
