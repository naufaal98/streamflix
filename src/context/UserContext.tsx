import { User } from 'data/user/user.type';
import React from 'react';
import UserService from 'data/user/user.service';

const initialState: User = {
  balance: 0,
  purchased_movies: [],
};

type UserContextType = {
  userData: User;
  syncUserContext: () => void;
};

export const UserContext = React.createContext<UserContextType>({
  userData: initialState,
} as UserContextType);

export const UserProvider: React.FC = ({ children }) => {
  const [userData, setUserData] = React.useState<User>(UserService.getLocalData);

  const syncUserContext = () => {
    setUserData(UserService.getLocalData);
  };

  return (
    <UserContext.Provider value={{ userData, syncUserContext }}>{children}</UserContext.Provider>
  );
};
