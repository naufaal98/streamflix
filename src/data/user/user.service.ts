import { Movie } from 'data/movie/movie.type';
import { User } from 'data/user/user.type';

const FREE_BALANCE = 100000;

const setInitialData = () => {
  const user: User = {
    balance: FREE_BALANCE,
    purchased_movies: [],
  };
  localStorage.setItem('user', JSON.stringify(user));
  return user;
};

const setLocalData = (user: User) => {
  try {
    localStorage.setItem('user', JSON.stringify(user));
  } catch (e) {
    // eslint-disable-next-line no-console
    console.error(e);
  }
};

const getLocalData = () => {
  try {
    if (!localStorage.getItem('user')) return setInitialData();
    const user = JSON.parse(localStorage.getItem('user')!) || [];
    return {
      balance: user.balance,
      purchased_movies: user.purchased_movies || [],
    };
  } catch (err) {
    return {
      balance: 0,
      purchased_movies: [],
    };
  }
};

const isMoviePurchased = (id: number) => {
  try {
    const userData = getLocalData();
    return !!userData!.purchased_movies.find((movie: Movie) => movie.id === id);
  } catch (err) {
    return false;
  }
};

export default { setLocalData, getLocalData, isMoviePurchased };
