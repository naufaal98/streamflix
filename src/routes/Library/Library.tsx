import MovieCard from 'components/MovieCard/MovieCard';
import { UserContext } from 'context/UserContext';
import { Movie } from 'data/movie/movie.type';
import React from 'react';
import styles from './Library.module.scss';

export default function Library() {
  const { userData } = React.useContext(UserContext);
  const movies = userData.purchased_movies;

  return (
    <>
      <h2>Library</h2>
      <div className={styles.MoviesGrid}>
        {userData.purchased_movies.map((movie: Movie) => (
          <div key={movie.id}>
            <MovieCard
              movie={movie}
              isPurchased={
                !!userData.purchased_movies.find((purchasedMovie) => movie.id === purchasedMovie.id)
              }
              key={movie.id}
            />
          </div>
        ))}
      </div>
      {movies.length === 0 && <p>Sorry, we couldn&apos;t find anything :(</p>}
    </>
  );
}
