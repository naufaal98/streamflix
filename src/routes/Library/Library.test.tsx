import { Movie } from 'data/movie/movie.type';
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { UserContext } from 'context/UserContext';
import { render, screen } from '@testing-library/react';
import { User } from 'data/user/user.type';
import Library from './Library';

const fakeMovies: Movie[] = [
  {
    id: 1,
    title: 'What a great Title',
    rating: 7,
    overview: 'Not bad',
    poster_path: 'somewhere',
    slug: 'what-a-great-title',
    price: 15000,
  },
];

const userData: User = {
  balance: 10000,
  purchased_movies: fakeMovies,
};

test(`renders the user's purchased movies`, () => {
  render(
    <Router>
      <UserContext.Provider value={{ userData, syncUserContext: jest.fn() }}>
        <Library />
      </UserContext.Provider>
    </Router>,
  );

  expect(screen.getByText(fakeMovies[0].title)).toBeInTheDocument();
});

test('renders feedback when purchased movies are empty', () => {
  render(
    <Router>
      <UserContext.Provider
        value={{ userData: { balance: 0, purchased_movies: [] }, syncUserContext: jest.fn() }}
      >
        <Library />
      </UserContext.Provider>
    </Router>,
  );

  expect(screen.getByRole('status')).toMatchSnapshot();
});
