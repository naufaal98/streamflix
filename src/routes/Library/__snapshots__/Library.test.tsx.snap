// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders feedback when purchased movies are empty 1`] = `
<p
  role="status"
>
  You haven't made any purchases yet. 
  <a
    href="/"
  >
    Start Explore movies
  </a>
</p>
`;
